### PR TITLE
don't close the client connection when the Public Header can't be parsed

### LINKED ISSUE #730: don't close the connection when receiving a packet with an un-parseable Public Header in the client

### PR DESCRIPTION
**Pull request by marten-seemann**
Fixes #730.

When receiving an unparseable packet with a spoofed remote address, we should not close the connection.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -211,22 +211,20 @@ func (c *client) listen() {
...
 	rcvTime := time.Now()
 
 	r := bytes.NewReader(packet)
 	hdr, err := ParsePublicHeader(r, protocol.PerspectiveServer)
 	if err != nil {
-		return qerr.Error(qerr.InvalidPacketHeader, err.Error())
+		utils.Errorf("error handling packet: %s", err.Error())
+		// drop this packet if we can't read parse the Public Header
+		// this packet could be malformed, or an injected packet with a spoofed remote address
+		return
```
Comment (marten-seemann) [inline client.go]: @lucas-clemente: Do we need to put back the packet buffer here?

Comment (lucas-clemente) [inline client.go]: We can put it back on every code path that doesn't enter the session. Putting it back is never required though, just a (small) matter of performance.

File: client.go [hunk 2]
```diff
@@ -211,22 +211,20 @@ func (c *client) listen() {
...
+func (c *client) handlePacket(remoteAddr net.Addr, packet []byte) {
 	rcvTime := time.Now()
 
 	r := bytes.NewReader(packet)
 	hdr, err := ParsePublicHeader(r, protocol.PerspectiveServer)
 	if err != nil {
-		return qerr.Error(qerr.InvalidPacketHeader, err.Error())
+		utils.Errorf("error handling packet: %s", err.Error())
+		// drop this packet if we can't read parse the Public Header
+		// this packet could be malformed, or an injected packet with a spoofed remote address
```
Comment (lucas-clemente) [inline client.go]: I don't understand this comment. How would ParsePublicHeader know it was a spoofed packet?

File: client.go [hunk 3]
```diff
@@ -211,22 +211,20 @@ func (c *client) listen() {
...
-func (c *client) handlePacket(remoteAddr net.Addr, packet []byte) error {
+func (c *client) handlePacket(remoteAddr net.Addr, packet []byte) {
 	rcvTime := time.Now()
 
 	r := bytes.NewReader(packet)
 	hdr, err := ParsePublicHeader(r, protocol.PerspectiveServer)
 	if err != nil {
-		return qerr.Error(qerr.InvalidPacketHeader, err.Error())
+		utils.Errorf("error handling packet: %s", err.Error())
+		// drop this packet if we can't read parse the Public Header
```
Comment (lucas-clemente) [inline client.go]: duplicate verb

File: client.go [hunk 4]
```diff
@@ -211,22 +211,20 @@ func (c *client) listen() {
...
 
-func (c *client) handlePacket(remoteAddr net.Addr, packet []byte) error {
+func (c *client) handlePacket(remoteAddr net.Addr, packet []byte) {
 	rcvTime := time.Now()
 
 	r := bytes.NewReader(packet)
 	hdr, err := ParsePublicHeader(r, protocol.PerspectiveServer)
 	if err != nil {
-		return qerr.Error(qerr.InvalidPacketHeader, err.Error())
+		utils.Errorf("error handling packet: %s", err.Error())
```
Comment (lucas-clemente) [inline client.go]: s/handling/parsing/

I'd include the remote addr in this message as well.

File: client.go [hunk 5]
```diff
@@ -239,21 +237,21 @@ func (c *client) handlePacket(remoteAddr net.Addr, packet []byte) error {
 		// otherwise this might be an attacker trying to inject a PUBLIC_RESET to kill the connection
 		if cr.Network() != remoteAddr.Network() || cr.String() != remoteAddr.String() || hdr.ConnectionID != c.connectionID {
```
Comment (lucas-clemente) [inline client.go]: Remark: it's trivial to just spoof the sender address, so the only really relevant thing here is the connection id, right?

File: client_test.go [hunk 6]
```diff
@@ -265,62 +257,53 @@ var _ = Describe("Client", func() {
...
 			})
 
-			It("errors if the version is supported by quic-go, but disabled by the quic.Config", func() {
-				v := protocol.SupportedVersions[1]
-				Expect(v).ToNot(Equal(cl.version))
-				Expect(config.Versions).ToNot(ContainElement(v))
-				err := cl.handlePacket(nil, composeVersionNegotiation(0x1337, []protocol.VersionNumber{v}))
-				Expect(err).To(MatchError(qerr.InvalidVersion))
-			})
+			// It("errors if the version is supported by quic-go, but disabled by the quic.Config", func() {
```
Comment (lucas-clemente) [inline client_test.go]: No commented code please. Either kill or set pending with an explanation.

File: client.go (hunk 2 above)
Comment (marten-seemann) [inline client.go]: It doesn't really. It could just be random junk arriving on our UDP socket.

File: client.go (hunk 5 above)
Comment (marten-seemann) [inline client.go]: Correct.

File: client_test.go (hunk 6 above)
Comment (marten-seemann) [inline client_test.go]: Oh, I forgot to reenable this test.

File: client.go (hunk 2 above)
Comment (lucas-clemente) [inline client.go]: Then I'd remove the second line of the comment.

**Pull request merged**